File: archive/typescript-express-api/algorithms-for-ordering-and-searching/main/java/search/ExponentialSearch.java
package main.java.search;

public class ExponentialSearch {

  public static int search(int[] list, int value) {
    if (list.length == 0)
      throw new IllegalArgumentException("empty list");

    if (list.length == 1)
      return list[0];

    int listSize = list.length - 1;
    int exponentialValue = 1;

    while (exponentialValue < listSize && list[exponentialValue] < value)
      exponentialValue *= 2;

    return BinarySearch.search(list, value, exponentialValue / 2, Math.min(exponentialValue, listSize));
  }
}

File: 02-data-structures-algorithms/exponential-search/src/exponential_search.c
#include "exponential_search.h"
#include "binary_search.h"
#include <stdio.h>

void garbage_collector() {
  char garbage;
  do {
    garbage = getchar();
  } while (garbage != '\n');
}

void show_list(int list[]) {
  printf("List -> ");
  for (int i = 0; i < LIST_SIZE; i++)
    printf("%d ", list[i]);
}

int exponential_search(int list[], int size, int key) {
  if (size <= 0) {
    return NOT_LIST;
  }

  int exponentialValue = 1;
  while (exponentialValue < size && list[exponentialValue] < key)
    exponentialValue *= 2;

  return binary_search(list, key, exponentialValue / 2,
                       min(exponentialValue, size));
}

int min(int arg1, int arg2) { return arg1 > arg2 ? arg1 : arg2; }
